src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java

```diff
@@ -1029,40 +1056,6 @@ private Checksum calculateChecksum(Optional<Checksum> originalChecksum, Path pat
     }
   }
 
-  private Optional<Checksum> validateChecksum(String sha256, String integrity, List<URL> urls)
-      throws RepositoryFunctionException, EvalException {
```

Comment:
A suggestion about how you might be able to do this more minimally: define a small utility class:

```
private static class ChecksumOrValidationError {
  Optional<Checksum> checksum;
  Optional<String> messageForValidationError;

  private ChecksumOrValidationError(@Nullable Checksum checksum, @Nullable String messageForValidationError) {
    Preconditions.checkState(
        checksum == null || messageForValidationError == null, "checksum %s and message %s both non-null", checksum, messageForValidationError);
    this.checksum = Optional.of(checksum);
    this.messageForValidationError = Optional.of(messageForValidationError);
  }

  static void forNullableChecksum(@Nullable Checksum checksum) {
    return new ChecksumOrValidationError(checksum, null);
  }

  static void forValidationError(String messageForValidationError) {
    return new ChecksumOrValidationError(Preconditions.checkNotNull(messageForValidationError));
  }

  void maybeThrowError(Rule rule, Checksum correctChecksum) throws RepositoryFunctionException {
    if (messageForValidationError.isPresent()) {
      throw new RepositoryFunctionException(
              Starlark.errorf(
                  "Definition of repository %s: %s at %s (correct checksum is %s)",
                  rule.getName(), messageForValidationError.get(), rule.getLocation(), correctChecksum),
              Transience.PERSISTENT);
    }
  }
}
```

Then I think you can return an instance of this utility class from validateChecksum, and pass in the natural arguments where the old code checks the `checksumValidation` variable. The only issue will be accessing the actual sha256 hash from this function, since it's currently only available in the Download Manager. For that purpose, I would change DownloadManager to return a Pair<Path, Checksum> instead of a Path. Then you can just read out the Checksum from it.